plutus-ledger/src/Ledger/TimeSlot.hs

```diff
@@ -1,45 +1,111 @@
+-- (2020-07-29T21:44:51Z) which is 1596059091 in POSIX time
+-- (number of seconds since 1970-01-01T00:00:00Z).
+beginOfTime :: Integer
+beginOfTime = 1596059091
 
 {-# INLINABLE slotRangeToPOSIXTimeRange #-}
--- | Convert a 'SlotRange' to 'POSIXTimeRange'
-slotRangeToPOSIXTimeRange :: SlotRange -> POSIXTimeRange
-slotRangeToPOSIXTimeRange sr = slotToPOSIXTime <$> sr
+-- | Convert a 'SlotRange' to a 'POSIXTimeRange' given a 'SlotConfig'. The
+-- resulting 'POSIXTimeRange' refers to the starting time of the lower bound of
+-- the 'SlotRange' and the ending time of the upper bound of the 'SlotRange'.
+slotRangeToPOSIXTimeRange :: SlotConfig -> SlotRange -> POSIXTimeRange
+slotRangeToPOSIXTimeRange sc sr =
+  let lbound = fmap (slotToBeginPOSIXTime sc) $ ivFrom sr
+      ubound = fmap (slotToEndPOSIXTime sc) $ ivTo sr
+   in Interval lbound ubound
 
-{-# INLINABLE slotToPOSIXTime #-}
--- | Convert a 'Slot to 'POSIXTime
-slotToPOSIXTime :: Slot -> POSIXTime
-slotToPOSIXTime (Slot n) = POSIXTime (n + shelleyLaunchDate)
+{-# INLINABLE slotToPOSIXTimeRange #-}
+-- | Convert a 'Slot' to a 'POSIXTimeRange' given a 'SlotConfig'. Each 'Slot'
+-- can be represented by an interval of time.
+slotToPOSIXTimeRange :: SlotConfig -> Slot -> POSIXTimeRange
+slotToPOSIXTimeRange sc slot =
+  interval (slotToBeginPOSIXTime sc slot) (slotToEndPOSIXTime sc slot)
+
+{-# INLINABLE slotToBeginPOSIXTime #-}
+-- | Get the starting 'POSIXTime' of a 'Slot' given a 'SlotConfig'.
+slotToBeginPOSIXTime :: SlotConfig -> Slot -> POSIXTime
+slotToBeginPOSIXTime SlotConfig{scSlotLength, scZeroSlotTime} (Slot n) =
+  let secondsAfterBegin = n * scSlotLength
+   in POSIXTime $ getPOSIXTime scZeroSlotTime + secondsAfterBegin
+
+{-# INLINABLE slotToEndPOSIXTime #-}
+-- | Get the ending 'POSIXTime' of a 'Slot' given a 'SlotConfig'.
+slotToEndPOSIXTime :: SlotConfig -> Slot -> POSIXTime
+slotToEndPOSIXTime sc@SlotConfig{scSlotLength} slot =
+  slotToBeginPOSIXTime sc slot + POSIXTime (scSlotLength - 1)
 
 {-# INLINABLE posixTimeRangeToSlotRange #-}
--- | Convert a 'POSIXTimeRange' to 'SlotRange'
-posixTimeRangeToSlotRange :: POSIXTimeRange -> SlotRange
-posixTimeRangeToSlotRange ptr = posixTimeToSlot <$> ptr
+-- | Convert a 'POSIXTimeRange' to 'SlotRange' given a 'SlotConfig'.
+posixTimeRangeToSlotRange :: SlotConfig -> POSIXTimeRange -> SlotRange
+posixTimeRangeToSlotRange sc = fmap (posixTimeToSlot sc)
 
 {-# INLINABLE posixTimeToSlot #-}
--- | Convert a 'POSIXTime' to 'Slot'
-posixTimeToSlot :: POSIXTime -> Slot
-posixTimeToSlot (POSIXTime t) = Slot (t - shelleyLaunchDate)
+-- | Convert a 'POSIXTime' to 'Slot' given a 'SlotConfig'.
+posixTimeToSlot :: SlotConfig -> POSIXTime -> Slot
+posixTimeToSlot SlotConfig{scSlotLength, scZeroSlotTime} (POSIXTime t) =
+  let timePassed = t - getPOSIXTime scZeroSlotTime
+      slotsPassed = divide timePassed scSlotLength
+  in Slot slotsPassed
+
+-- | Get the current slot number
+currentSlot :: SlotConfig -> IO Slot
```

Comment:
This seems kind of dangerous, do we use this?

Reply:
Yeah, it's used in `Cardano.Node.Mock` in `plutus-pab`. That function came from there originally.

Reply:
Also used in `Cardano.Protocol.Socket.Client`